galerautils/src/gu_enc_mmap.cpp

```diff
@@ -0,0 +1,628 @@
+    }
+    S_DEBUG("vpage -> ppage mappings end\n");
+}
+
+
+// Helper methods
+unsigned char* EncMMap::page_start(unsigned long long pageNo) const {
+    return base_ + pageSize_ * pageNo;
+}
+
+unsigned char* EncMMap::page_start(unsigned char* addr) const {
+    unsigned long long addr_u = ptr2ull(addr);
+    unsigned long long page_start = (addr_u / pageSize_) * pageSize_;
+    return reinterpret_cast<unsigned char*>(page_start);
+}
+
+size_t EncMMap::page_number(unsigned char* addr) const {
+    unsigned long long offset = ptr2ull(addr) - ptr2ull(base_);
+    size_t pageNo = offset / pageSize_;
+    return pageNo;
+}
+
+
+// encrption / decryption
+void EncMMap::encrypt(unsigned char* dst, unsigned char* src, size_t size, size_t pageNumber) const {
+    // the last page may be not full
+    size = IS_LAST_PAGE(pageNumber) ? lastPageSize_ : size;
+#if REAL_ENCRYPTION
+    size_t pageStartOffset = pageNumber * pageSize_;
+    size_t unencryptedSize = 0;
+
+    if (gu_unlikely(pageStartOffset < encryptionStartOffset_)) {
+        unencryptedSize = std::min(size, encryptionStartOffset_);
+        memcpy(dst, src, unencryptedSize);
+        dst += unencryptedSize;
+        src += unencryptedSize;
+    }
+
+    int encryptedSize = size - unencryptedSize;
+    if(encryptedSize > 0) {
+      encryptor_.set_stream_offset(pageStartOffset + unencryptedSize);
+      encryptor_.encrypt(dst, src, size - unencryptedSize);
+    }
+#else
+    memcpy(dst, src, size);
+#endif
+}
+
+void EncMMap::decrypt(unsigned char* dst, unsigned char* src, size_t size, size_t pageNumber) const {
+    // the last page may be not full
+    size = IS_LAST_PAGE(pageNumber) ? lastPageSize_ : size;
+#if REAL_ENCRYPTION
+    size_t pageStartOffset = pageNumber * pageSize_;
+    size_t unencryptedSize = 0;
+
+    if (gu_unlikely(pageStartOffset < encryptionStartOffset_)) {
+        unencryptedSize = std::min(size, encryptionStartOffset_);
+        memcpy(dst, src, unencryptedSize);
+        dst += unencryptedSize;
+        src += unencryptedSize;
+    }
+
+    int encryptedSize = size - unencryptedSize;
+    if(encryptedSize > 0) {
+        decryptor_.set_stream_offset(pageStartOffset + unencryptedSize);
+        decryptor_.decrypt(dst, src, size - unencryptedSize);
+    }
+#else
+    memcpy(dst, src, size);
+#endif
+}
+
+
+// signal handler install and dispatch
+static std::once_flag signal_handler_once;
+static struct sigaction oldsigact;
+
+void signal_handler(int sig, siginfo_t* info, void* ctx) {
+    // If someone is accessing encMMaps, just bail out without even trying
+    // to find our object.
+    // It can happen in the following situations:
+    // 1. (rare) The client in registering/deregistering new EncMMap object
+    // 2. (more probable but still not so often) Two signal handlers are called
+    //    simultaneously (2 threads). In such a case the 2nd one will retry
+    //    in a while.
+    if (encMmapsLock.test_and_set(std::memory_order_acquire)) {
```

Comment:
We have to use test_and_set because we can't use mutexes in signal handlers?

In that case, it would be useful to add a RAII helper to this, similar to scoped_lock.

Reply:
Yes, this is the reason.
As for RAII helper, could you give the example how it would work?
scoped_lock locks in its constructor. There is no possibility to NOT lock. It waits until the lock is possible, then goes on, so once the object is created, it is locked.
Here we are locking or not (something like try-lock). 
How to RAII it? Do you mean something like:
```
std::unique_lock<std::mutex> lock(_mutex, std::try_to_lock);
if(!lock.owns_lock()){
    // mutex wasn't locked. Handle it.
}
```
Is it worth it? Please check `addEncMMap` and `delEncMMap`. I'm not sure it will make it any simpler/safer

Reply:
Yes, it would make sense only for addEncMMap and delEncMMap. I know those are only small functions, so maybe its not worth it.

Reply:
This part has been reworked.